File: 990.等式方程的可满足性.py
#
# @lc app=leetcode.cn id=990 lang=python3
#
# [990] 等式方程的可满足性
#

# @lc code=start
from typing import List, Dict

class DisjointSet:
    def __init__(self, _size) -> None:
        self.size = _size
        self.data = [-1] * _size
    
    def query(self, val: int) -> int:
        if self.data[val] < 0: return val
        self.data[val] = self.query(self.data[val])
        return self.data[val]

    def merge(self, a: int, b: int) -> None:
        root_a, root_b = self.query(a), self.query(b)
        if root_a == root_b: return
        if self.data[root_a] < self.data[root_b]: root_a, root_b = root_b, root_a
        self.data[root_a], self.data[root_b] = self.data[root_a] + self.data[root_b], root_a

class Solution:
    def equationsPossible(self, equations: List[str]) -> bool:
        idxer: Dict[str, int] = {}
        for eq in equations:
            if eq[0] not in idxer: idxer[eq[0]] = len(idxer)
            if eq[3] not in idxer: idxer[eq[3]] = len(idxer)
        djs = DisjointSet(len(idxer))
        for eq in equations:
            if eq[1] == '!': continue
            djs.merge(idxer[eq[0]], idxer[eq[3]])
        for eq in equations:
            if eq[1] == '!' and djs.query(idxer[eq[0]]) == djs.query(idxer[eq[3]]):
                return False
        return True
# @lc code=end

File: 990.等式方程的可满足性.cpp
/*
 * @lc app=leetcode.cn id=990 lang=cpp
 *
 * [990] 等式方程的可满足性
 */

// @lc code=start
#include<vector>
#include<string>
#include<unordered_map>
using namespace std;

class DisjointSet{
public:
    int size;
    vector<int> data;
    DisjointSet(int _size): size(_size), data(_size, -1) {}
    int query(int val){
        if(data[val] < 0) return val;
        return data[val] = query(data[val]);
    }
    void merge(int a, int b){
        int root_a = query(a), root_b = query(b);
        if(root_a == root_b) return;
        if(data[root_a] > data[root_b]) swap(root_a, root_b);
        auto tmp = data[root_b];
        data[root_b] = root_a;
        data[root_a] += tmp;
    }
};

class Solution {
public:
    bool equationsPossible(const vector<string>& equations) {
        unordered_map<char, int> char2idx;
        for(auto& eq: equations){
            if(!char2idx.count(eq.front())) char2idx[eq.front()] = char2idx.size();
            if(!char2idx.count(eq.back())) char2idx[eq.back()] = char2idx.size();
        }
        DisjointSet djs(char2idx.size());
        for(auto& eq: equations){
            if(eq[1] == '=') djs.merge(char2idx[eq.front()], char2idx[eq.back()]);
        }
        for(auto& eq: equations){
            if(eq[1] == '!' && djs.query(char2idx[eq.front()]) == djs.query(char2idx[eq.back()])) return false;
        }
        return true;
    }
};
// @lc code=end

#include<iostream>
int main(){
    cout << Solution().equationsPossible({"a==b", "b!=a"}) << '\n'
         << Solution().equationsPossible({"b==a", "a==b"}) << '\n'
         << Solution().equationsPossible({"a==b", "b==c", "a==c"}) << '\n'
         << Solution().equationsPossible({"a==b", "b!=c", "c==a"}) << '\n'
         << Solution().equationsPossible({"c==c", "b==d", "x!=z"}) << '\n';
    return 0;
}